Initial implementation for ESLint Formatter Todo

renatoi opened the pull request:
Implements the [todo spec](https://github.com/ember-template-lint/ember-template-lint-todo-utils/blob/master/docs/spec.md) for ESLint as a custom formatter to allow errors to pass linting as warnings as long as they are in the todo list. The todo list is also generated when the flag `UPDATE_TODO` is not set to `"1"`.

scalvert commented on src/index.ts:
```diff
@@ -0,0 +1,4 @@
+import { formatResults } from "./formatter";
+
+// special TS syntax: https://github.com/microsoft/TypeScript/issues/2719
+export = formatResults;
```
Are you doing this because eslint expects a formatter to have a top level function export that is the reporter function?

renatoi replied:
Correct, otherwise it complains that it got an object instead of a function.

scalvert commented on src/formatter.ts:
```diff
@@ -0,0 +1,36 @@
+}
+
+async function report(results: ESLint.LintResult[], todoMap?: Map<string, TodoData>) {
+  mutateTodoErrorsToWarnings(results, todoMap);
+  const eslint = new ESLint({});
+  const defaultFormatter = await eslint.loadFormatter('stylish');
+  console.log(defaultFormatter.format(results));
```
Could we add some breathing room in the form of blank lines here? It helps for readability.

scalvert commented on src/formatter.ts:
```diff
@@ -0,0 +1,36 @@
+}
+
+async function formatResultsAsync(results: ESLint.LintResult[]): Promise<void> {
+  const UPDATE_TODO: string = process.env.UPDATE_TODO || '0';
+  const basePath = process.env.ESLINT_TODO_DIR || process.cwd();
+  if (UPDATE_TODO == '1') {
+    await generateTodoFiles(basePath, results);
+    report(results);
+  } else {
+    const todoDir = getTodoDirPath(basePath);
+    if (fs.existsSync(todoDir)) {
+      const todoMap = await readTodoFiles(todoDir);
+      report(results, todoMap);
+    } else {
+      report(results);
+    }
+  }
```
Could we add some breathing room in the form of blank lines here? It helps for readability.

renatoi replied:
Done. I also changed the code to return early so nesting is reduced to improve readability.

scalvert commented on src/mutate-errors-to-warnings.ts:
```diff
@@ -0,0 +1,32 @@
+import { generateFileName, _buildTodoDatum } from '@ember-template-lint/todo-utils';
+import { TodoData } from '@ember-template-lint/todo-utils/lib/types';
```
Let's make a note that we need to export the types from that package, as 'digging in' to a private module is brittle and should be avoided.

renatoi replied:
I added a `TODO` comment to note this.

renatoi replied:
Also created https://github.com/ember-template-lint/ember-template-lint-todo-utils/issues/18

scalvert commented on src/mutate-errors-to-warnings.ts:
```diff
@@ -0,0 +1,32 @@
+): void {
+  results.forEach((result: ESLint.LintResult) => {
+    result.messages.forEach((message) => {
+      if (message.severity === 2) {
+        const todoDatum = _buildTodoDatum(result, message);
+        const todoHash = generateFileName(todoDatum);
+        if (!todoMap || todoMap.get(todoHash)) {
+          message.severity = 1;
+          result.errorCount -= 1;
+          result.warningCount += 1;
+          if (message.fix) {
+            result.fixableErrorCount -= 1;
+            result.fixableWarningCount += 1;
+          }
+        }
+      }
```
Could we add some breathing room in the form of blank lines here? It helps for readability.

renatoi replied:
Done. I also changed the code to return early so nesting is reduced to improve readability.

renatoi commented on README.md:
````diff
@@ -1 +1,23 @@
+$ UPDATE_TODO=1 eslint --format eslint-formatter-todo
+```
+
+This command will transform all reported ERRORS to WARNINGS and will generate files inside the `.lint-todo` directory to track them.
````
So `-1` is not an actual valid value for ESLint. Valid values are `0` (off), `1` (warning), and `2` (error). Going with `-1` has the potential of getting undefined behavior in the future while it also doesn't let me set it in typescript since `@types` restrict severity to only be those 3 values. I could set it to `0` though, and force the formatter to show them as `todo`, thoughts?

scalvert replied:
We're aware that it's not a valid value for Severity, but that can be fixed through type extension. 

It's not undefined behavior, because that behavior is only relevant to us for the purposes of this functionality. We're not persisting the severity anywhere, and we're simply changing the severity for the rendered output. 

Ultimately, if this feature pans out we'd like to propose it as a feature in `eslint` proper. It's already getting build into `ember-template-lint`.

renatoi commented on __tests__/formatter-test.ts:
```diff
@@ -0,0 +1,47 @@
+import { ESLint } from 'eslint';
+import fs from 'fs';
+import path from 'path';
+import tmp from 'tmp';
```
Done.

renatoi commented on __tests__/formatter-test.ts:
```diff
@@ -0,0 +1,47 @@
+  it('SHOULD NOT generate a TODO dir with todo files when UPDATE_TODO is set to 0', async () => {
+    // set the variables
+    process.env.UPDATE_TODO = '0';
+    const results = require('./__fixtures__/eslint-with-errors.json') as ESLint.LintResult[];
```
So the problem is that the code mutates this object tree which affects other tests. The reason why I'm requiring and also resetting module the module on `beforeEach` is because I want a brand new copy for every test.

renatoi replied:
Ok. I fixed this, I figured using `resetModules` with `require` in tests is brittle and prone to problems, so I decided to just read the JSON directly instead, this way we won't have to do any of those weird calls.

renatoi commented on __tests__/mutate-errors-to-warnings-test.ts:
```diff
@@ -0,0 +1,60 @@
+import { mutateTodoErrorsToWarnings } from '../src/mutate-errors-to-warnings';
+
+beforeEach(() => {
+  jest.resetModules();
```
As I mentioned above, this is to ensure that every copy of the JSON we require in the tests are brand new copies and not the mutated ones.

renatoi replied:
Ok. I fixed this, I figured using `resetModules` with `require` in tests is brittle and prone to problems, so I decided to just read the JSON directly instead, this way we won't have to do any of those weird calls.

renatoi commented on __tests__/mutate-errors-to-warnings-test.ts:
```diff
@@ -0,0 +1,60 @@
+
+describe('mutate-errors-to-warnings', () => {
+  it('changes all errors to warnings when a todoMap IS NOT passed', () => {
+    const results = require('./__fixtures__/eslint-with-errors.json') as ESLint.LintResult[];
```
Same as above.

renatoi commented on __tests__/mutate-errors-to-warnings-test.ts:
```diff
@@ -0,0 +1,60 @@
+        result.fixableWarningCount + result.fixableErrorCount
+      );
+      result.messages.forEach((message) => {
+        message.severity != 2;
```
Removed.

renatoi commented on jest.config.js:
```diff
@@ -0,0 +1,18 @@
+      statements: 100,
+    },
+  },
+  testPathIgnorePatterns: ['/__fixtures__/', '/__utils__/'],
```
Yep good catch. I was using one but removed it later.

renatoi commented:
Addressed comments in another PR (https://github.com/scalvert/eslint-formatter-todo/pull/3), closing this one. A new PR was created so the change is now part of the same repo instead of a fork.
